packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp

```diff
@@ -0,0 +1,301 @@
+//@HEADER
+#ifndef _ZOLTAN2_ICESHEET_HPP_
+#define _ZOLTAN2_ICESHEET_HPP_
+#include<unordered_set>
+#include<utility>
+#include<vector>
+#include<Zoltan2_GraphModel.hpp>
+#include<Zoltan2_Algorithm.hpp>
+#include<Zoltan2_PartitioningSolution.hpp>
+#include<Zoltan2_Util.hpp>
+#include<Zoltan2_TPLTraits.hpp>
+#include<Zoltan2_VtxLabel.hpp>
+#include<Zoltan2_AlltoAll.hpp>
+
+namespace Zoltan2{
+  template <typename Adapter>
+  class IceProp : public Algorithm<Adapter> {
+    public:
+      typedef typename Adapter::base_adapter_t base_adapter_t;
+      typedef typename Adapter::lno_t lno_t;
+      typedef typename Adapter::gno_t gno_t;
+      typedef typename Adapter::offset_t offset_t;
+      typedef typename Adapter::scalar_t scalar_t;
+      typedef typename Adapter::part_t part_t;
+      typedef typename Adapter::user_t user_t;
+      typedef typename Adapter::userCoord_t userCoord_t;
+      typedef Tpetra::Map<lno_t, gno_t> map_t;
+      
+      //Arguments: problemComm is the communicator we will use for the propagation
+      //           adapter is the graph adapter that represents the ice mesh's bottom layer
+      //           basalFriction is the array of basalFriction values. Nonzero indicates grounded
+      //           boundary_edges is the array of edges on the current processor that are on the
+      //                          boundary, using global identifiers
+      //           num_boundary_edges is the number of boundary edges there are
+      IceProp(const RCP<const Comm<int> > &problemComm__,
+	      const RCP<const GraphAdapter<user_t,userCoord_t> > &adapter__, 
+	      bool* basalFriction, gno_t* boundary_edges__, int num_boundary_edges__):
+         adapter(adapter__), grounding_flags(basalFriction), boundary_edges(boundary_edges__),
+         num_boundary_edges(num_boundary_edges__), problemComm(problemComm__)
+      {
+
+	env = rcp(new Environment(problemComm));
+        modelFlag_t flags;
+	flags.reset();
+
+	///build the graph model from the GraphAdapter.	 
+	buildModel(flags);
+      }
+   
+      //This function returns a number of flags consistent with the 
```

Comment:
Since this is the main interface for user, it needs better and doxygenized documentation of outputs.  What is returned?  How does one interpret the flags' values?  Etc.

Reply:
I have added better documentation.